### PR TITLE
feat: dcp async checkpoint integration

### LINKED ISSUE #21557: Revamp Async Checkpointing to Use PyTorch Distributed Checkpoint (DCP) async_save
### Description & Motivation

PyTorch Lightning’s current async checkpointing implementation predates PyTorch’s Distributed Checkpoint (DCP) API and feels outdated.

This issue proposes evaluating and migrating Lightning’s async checkpoint logic to leverage `torch.distributed.checkpoint` (DCP), specifically `async_save`, to:

* Align with upstream PyTorch checkpointing APIs
* Improve robustness and maintainability
* Better support distributed and sharded training setups
* Reduce custom logic that duplicates upstream functionality

### Pitch

Use [PyTorch DCP's async_save](https://docs.pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict_saver.async_save)

### Alternatives

_No response_

### Additional context

https://docs.pytorch.org/docs/stable/distributed.checkpoint.html#distributed-checkpoint-torch-distributed-checkpoint

cc @lantiga

### PR DESCRIPTION
**Pull request by deependujha**
## What does this PR do?

Fixes #21557

Integrates `DCP async checkpointing` into `PyTorch Lightning`, enabling checkpoints to be written asynchronously to reduce training-loop blocking and improve throughput, especially for large models and slower storage backends.

```python
# with fabric
from lightning.fabric.plugins import DistributedAsyncCheckpointIO
fabric = Fabric(plugins=[DistributedAsyncCheckpointIO()])

# with trainer
from lightning.pytorch.plugins import DistributedAsyncCheckpointIO
trainer = L.Trainer(plugins=[DistributedAsyncCheckpointIO()])
```

### Minimal Benchmark (`Async vs Sync`) (T4 4 gpu machine)

Screenshot 2026-02-15 at 6 25 56 PM

- Smaller model: ~2.7× faster saves
- Larger model: ~5× faster saves
- Expect larger gains as model size increases (better overlap of compute + IO)

---

### Minimal example with fabric

<details>

<summary>Saving Checkpoint</summary>

```python
import torch
from torch import nn
from torch.utils.data import DataLoader, TensorDataset
from lightning.fabric import Fabric
from lightning.fabric.utilities import AttributeDict
from lightning.fabric.plugins import DistributedAsyncCheckpointIO


# Simple model
class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(10, 2)

    def forward(self, x):
        return self.layer(x)


# Training loop
def train_one_epoch(fabric, model, optimizer, dataloader, loss_fn, epoch):
    model.train()
    for step, (x, y) in enumerate(dataloader):
        x, y = fabric.to_device((x, y))
        optimizer.zero_grad()
        preds = model(x)
        loss = loss_fn(preds, y)
        fabric.backward(loss)
        optimizer.step()

        # log step loss
        fabric.log("train_loss", loss.item(), step=epoch * len(dataloader) + step)
        print(f"Epoch {epoch} Step {step} Loss {loss.item():.4f}")


def main():
    # Dummy dataset
    x = torch.randn(100, 10)
    y = torch.randint(0, 2, (100,))
    dataset = TensorDataset(x, y)
    loader = DataLoader(dataset, batch_size=16)

    async_checkpoint_io = DistributedAsyncCheckpointIO()

    fabric = Fabric(plugins=[async_checkpoint_io])
    fabric.launch()

    # Setup
    model = SimpleModel()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    loss_fn = nn.CrossEntropyLoss()
    model, optimizer, loader = fabric.setup(model, optimizer, loader)

    # Run training
    for epoch in range(5):
        train_one_epoch(fabric, model, optimizer, loader, loss_fn, epoch)
        state = AttributeDict(model=model, optimizer=optimizer, iteration=epoch)
        fabric.save(f"checkpoint/checkpoint_epoch_{epoch}.pt", state)


if __name__ == "__main__":
    main()
```

</details>

<details>

<summary>Loading Checkpoint</summary>

```python
import torch
from torch import nn
from torch.utils.data import DataLoader, TensorDataset
from lightning.fabric import Fabric
from lightning.fabric.utilities import AttributeDict
from lightning.fabric.plugins.io.distributed_async_io import DistributedAsyncCheckpointIO


# Simple model
class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(10, 2)

    def forward(self, x):
        return self.layer(x)


# Training loop
def train_one_epoch(fabric, model, optimizer, dataloader, loss_fn, epoch):
    model.train()
    for step, (x, y) in enumerate(dataloader):
        x, y = fabric.to_device((x, y))
        optimizer.zero_grad()
        preds = model(x)
        loss = loss_fn(preds, y)
        fabric.backward(loss)
        optimizer.step()

        print(f"{fabric.global_rank=}; Epoch {epoch} Step {step} Loss {loss.item():.4f}")


def main():
    # Dummy dataset
    x = torch.randn(100, 10)
    y = torch.randint(0, 2, (100,))
    dataset = TensorDataset(x, y)
    loader = DataLoader(dataset, batch_size=16)

    async_checkpoint_io = DistributedAsyncCheckpointIO()
    plugins = []
    plugins.append(async_checkpoint_io)

    fabric = Fabric(plugins=plugins)
    fabric.launch()

    # Setup
    model = SimpleModel()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    loss_fn = nn.CrossEntropyLoss()
    model, optimizer, loader = fabric.setup(model, optimizer, loader)

    state = AttributeDict(model=model, optimizer=optimizer, iteration=99)
    print(f"state before loading checkpoint: {state}")
    fabric.load("checkpoint/checkpoint_epoch_4.pt", state)
    print(f"state after loading checkpoint: {state}")

    
    # check output of input
    simple_input = torch.arange(10).float().unsqueeze(0).to(fabric.device)
    with torch.no_grad():
        out = model(simple_input)
    print(f"model output: {out}")
    print(f"{state['iteration']=}")


if __name__ == "__main__":
    main()
```

</details>

### Minimal example with Trainer

<details>

<summary>Saving Checkpoint</summary>

```python
import torch
import lightning as L
from lightning.pytorch.plugins import DistributedAsyncCheckpointIO

# -- dataset --
x = torch.randn(100,4)         # random dataset of size 100, containing 4 features
y = torch.randint(0,2, (100,)) # random labels: {0, 1} (2 classes)

dataset = torch.utils.data.TensorDataset(x, y)
dataloader = torch.utils.data.DataLoader(dataset, batch_size = 4, drop_last=True)


# -- lightning module --

class SimpleModel(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(4, 8)
        self.act = torch.nn.ReLU()
        self.out = torch.nn.Linear(8,2) # number of classes 2

    def forward(self, x):
        return self.out(self.act(self.layer(x))) 

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_pred = self(x)
        loss = torch.nn.functional.cross_entropy(y_pred, y)
        self.log("loss", loss)
        return loss

    def configure_optimizers(self):
        opt = torch.optim.Adam(self.parameters(),lr = 1e-3)
        return opt

def main():
    model = SimpleModel()
    dist_async_ckpt = DistributedAsyncCheckpointIO()

    trainer = L.Trainer(max_epochs=10, log_every_n_steps=1, plugins=[dist_async_ckpt])

    ckpt_path = "lightning_logs/version_6/checkpoints/epoch=9-step=250.ckpt"
    # ckpt_path = None
    trainer.fit(model=model, train_dataloaders=dataloader, ckpt_path=ckpt_path)

if __name__ == "__main__":
    main()
```

</details>

<details>

<summary>Loading checkpoint </summary>

```python
import torch
import lightning as L
from lightning.pytorch.plugins import DistributedAsyncCheckpointIO

# -- dataset --
x = torch.randn(100,4)         # random dataset of size 100, containing 4 features
y = torch.randint(0,2, (100,)) # random labels: {0, 1} (2 classes)

dataset = torch.utils.data.TensorDataset(x, y)
dataloader = torch.utils.data.DataLoader(dataset, batch_size = 4, drop_last=True)


# -- lightning module --

class SimpleModel(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(4, 8)
        self.act = torch.nn.ReLU()
        self.out = torch.nn.Linear(8,2) # number of classes 2

    def forward(self, x):
        return self.out(self.act(self.layer(x))) 

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_pred = self(x)
        loss = torch.nn.functional.cross_entropy(y_pred, y)
        self.log("loss", loss)
        return loss

    def configure_optimizers(self):
        opt = torch.optim.Adam(self.parameters(),lr = 1e-3)
        return opt

def main():
    model = SimpleModel()
    dist_async_ckpt = DistributedAsyncCheckpointIO()

    trainer = L.Trainer(max_epochs=20, log_every_n_steps=1, plugins=[dist_async_ckpt])

    ckpt_path = "lightning_logs/version_0/checkpoints/epoch=9-step=250.ckpt"
    trainer.fit(model=model, train_dataloaders=dataloader, ckpt_path=ckpt_path)
    

if __name__ == "__main__":
    main()
```

</details>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21515.org.readthedocs.build/en/21515/

### CODE REVIEW

**Comment by deependujha** [comment]
next pr todos:

- use a queue to pipeline checkpoint saves. (update: done ✅)
```python
self.checkpoint_future_queue: Optional[Future] = queue.Queue(maxsize=2)
```

- verify for remote checkpoint dir

File: src/lightning/fabric/plugins/io/distributed_async_io.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    Notes:
+        - Loading is in-place: caller must provide a pre-allocated `state_dict`
+
+    """
+
+    def __init__(
+        self,
+        checkpointer_type: Optional[CHECKPOINTER_TYPE] = None,
+        no_dist: bool = True,
```
Comment (tchaton) [inline src/lightning/fabric/plugins/io/distributed_async_io.py]: Maybe `sharded_checkpoints = False` might be more intuitive. But also, this would require to be configured by the strategy.

File: src/lightning/fabric/plugins/io/distributed_async_io.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    @override
+    def save_checkpoint(
+        self, checkpoint: dict[str, Any], path: _PATH, storage_options: Optional[dict[str, Any]] = None
+    ) -> None:
+        if storage_options is not None:
+            raise TypeError(
+                "`storage_options` is not supported by AsyncCheckpointIO. Implement a custom CheckpointIO if needed."
+            )
+
+        self._wait()
```
Comment (tchaton) [inline src/lightning/fabric/plugins/io/distributed_async_io.py]: Do we need to wait there, can't we have multiple future being created ?

Comment (deependujha) [reply src/lightning/fabric/plugins/io/distributed_async_io.py]: I didn't do this for the first PR. But added it as todo: https://github.com/Lightning-AI/pytorch-lightning/pull/21515#issuecomment-3901941328

should i add this in this pr?

Comment (deependujha) [reply src/lightning/fabric/plugins/io/distributed_async_io.py]: done.
